### PR TITLE
Fix problem with burning tokens

### PR DESCRIPTION
**Pull request by UlfNorell**
tl;dr The constraint solver treated burn constraints as negative inputs, forcing it to add an extra pair of input/output to make up the "required spend" of zero tokens.

Transactions are constructed in three phases:

1. `Ledger.Constraints.OffChain.processContraints`:
   The constraints are traversed computing inputs, outputs and forges, and tracking two quantities: valueSpentActual, and valueSpentRequired.

2. `Ledger.Constraints.OffChain.addMissingValueSpent`:
   The positive part of `valueSpentRequired - valueSpentActual` is added as outputs to the pubkey signing the transaction.

3. `Language.Plutus.Contract.Wallet.balanceWallet`:
   The transaction is balanced, adding inputs and outputs to/from the wallet to make the sum of inputs, outputs and forges zero.

So what counts as valueSpentActual and valueSpentRequired in step 1?

Outside forging, `valueSpentActual` is increased every time an input is added to the transaction (`MustSpendPubKeyOutput` and `MustSpendScriptOutput`). Forging (`MustForgeValue`) a positive number of tokens is equivalent to spending an output, so we add these to `valueSpentActual` as well.

When adding an output (`MustPayToPubKey`/`MustPayToOtherScript`) the corresponding amount is subtracted from `valueSpentRequired`. In addition, `valueSpentRequired` is increased for `MustSpendValue` constraints. Since buring tokens (forging a negative amount of tokens) is a form of output we should subtract the number of tokens burned from `valueSpentRequired` in this case (**previously they were incorrectly subtracted from `valueSpentActual`**).

Some observations (not addressed by this PR):

- We only ever look at the difference between valueSpentActual and valueSpentRequired, so it would suffice to track a single Value representing the difference.
- Inputs are always added to valueSpentActual and outputs subtracted from valueSpentRequired, so we could skip this and add up inputs and outputs in step 2, leaving only the `valueSpentRequired` increases from `MustSpendValue`.

Pre-submit checklist:
- Branch
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [ ] ~~Relevant tickets are mentioned in commit messages~~
- PR
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested

Pre-merge checklist:
- [ ] Someone approved it
- [ ] Commits have useful messages
- [ ] Review clarifications made it into the code
- [ ] History is moderately tidy; or going to squash-merge

### CODE REVIEW

**Comment by j-mueller** [comment]
> it would suffice to track a single Value representing the difference.
> we could skip this and add up inputs and outputs in step 2

Thanks, yes those two changes will make it much easier to understand what's going on. I'll do that today so it doesn't get lost.

**Review by j-mueller — APPROVED** [review]
